src/module_4/push_exploration_non_linear.md

```diff
@@ -0,0 +1,2005 @@
+    use_label_encoder=False,
+    eval_metric='aucpr'
+)
+
+high_estimator_xgb_model.fit(X_train, y_train)
+y_pred_high_estimator_xgb_model = high_estimator_xgb_model.predict_proba(X_val)[:, 1]
+
+# Best XGBoost model based on the RandomizedSearchCV results
+deeper__xgb_model = xgb.XGBClassifier(
+    n_estimators=50,
+    learning_rate=0.05,
+    max_depth=6,
+    reg_lambda=0.5,  # 'lambda' is renamed to 'reg_lambda' in XGBClassifier
+    objective='binary:logistic',
+    use_label_encoder=False,
+    eval_metric='aucpr'
+)
+
+deeper__xgb_model.fit(X_train, y_train)
+y_pred_deeper__xgb_model = deeper__xgb_model.predict_proba(X_val)[:, 1]
+
+# Best XGBoost model based on the RandomizedSearchCV results
+high_deeper__xgb_model = xgb.XGBClassifier(
+    n_estimators=250,
+    learning_rate=0.05,
+    max_depth=6,
+    reg_lambda=0.5,  # 'lambda' is renamed to 'reg_lambda' in XGBClassifier
+    objective='binary:logistic',
+    use_label_encoder=False,
+    eval_metric='aucpr'
+)
+
+high_deeper__xgb_model.fit(X_train, y_train)
+y_pred_high_deeper__xgb_model = high_deeper__xgb_model.predict_proba(X_val)[:, 1]
+
+# Evaluate
+generate_evaluation_curves_multi(
+    model_names=["LogisticRegression"] + ["GBT(n_trees=100, depth=3)"] + ["XGBoost (n_trees=50, depth=3, lr=0.05)"] + ["High-estimator"] + ["Deeper"]+ ["High&Deeper"],
+    y_preds=[y_pred_lr, y_pred_gbt,y_pred_best_xgb, y_pred_high_estimator_xgb_model, y_pred_deeper__xgb_model, y_pred_high_deeper__xgb_model ],
+    y_test=y_val
+)
+```
+
+
+    
+![png](push_exploration_non_linear_files/push_exploration_non_linear_52_0.png)
+    
+
+
+
+```python
+plot_calibration_curve(
+    y_val,
+    [y_pred_best_xgb, y_pred_high_estimator_xgb_model, y_pred_deeper__xgb_model, y_pred_high_deeper__xgb_model],
+    ["XGBoost (n_trees=50, depth=3, lr=0.05)", "XGB but n_trees=250", "XGB but depth=6", "XGB (n_trees=250, depth=6)"]
+)
+
+
+```
+
+
+    
+![png](push_exploration_non_linear_files/push_exploration_non_linear_53_0.png)
+    
+
+
+Indeed they do, my gut-feeling it is that we are providing more complexity to the model, hence letting it decide with more robustness for positive outcomes (although they are not our best model as we saw previously).
+
+Having said that, let's move on calibrating our chosen model (catboost). Since we made use of pools, we need to extract the data from there in order to pass them into scikit libraries.
+
+
+```python
+from sklearn.calibration import CalibratedClassifierCV, calibration_curve
+from sklearn.preprocessing import OneHotEncoder
+
+def calibrate_model(model, X_train, y_train, method='sigmoid'):
+    # Fit the calibrator - note that we do not need a Pool object here
+    calibrated = CalibratedClassifierCV(model, method=method, cv='prefit')
+    calibrated.fit(X_train, y_train)
+    return calibrated
+
+# Assuming X_val is your validation feature matrix and y_val are the validation labels:
+calibrated_catboost = calibrate_model(catboost_model, X_val, y_val)
```

Comment:
you should always calibrate the model in train! otherwise you can do overfitting without noticing (does the calibration generalizes to other sets)